### ISSUE TITLE
Incorrect `torch.einsum` calculation with 1D FloatTensor

### ISSUE THREAD

**Issue by junhsss** [pytorch/pytorch#62218]
## 🐛 Bug

`torch.einsum` gives incorrect result when implementing an inner product between 1D FloatTensor. I suppose this problem is related to Apple M1 since it gives consistent results on other machines.

## To Reproduce

```python
import torch
x = torch.tensor([1., 2.], dtype=torch.float32)
print("float32 : ", torch.einsum('i, i -> ', x, x)) 

y = torch.tensor([1., 2.], dtype=torch.float64)
print("float64 : ", torch.einsum('i, i -> ', y, y)) 

z = torch.tensor([1, 2])
print("int : ", torch.einsum('i, i -> ', z, z))
```

This gives

```
float32 :  tensor(0.)
float64 :  tensor(5., dtype=torch.float64)
int :  tensor(5)
```

## Environment

PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.4 (arm64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.11)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.6 (default, Jun 28 2021, 19:24:41)  [Clang 12.0.5 (clang-1205.0.22.9)] (64-bit runtime)
Python platform: macOS-11.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.1
[pip3] torch==1.9.0
[conda] Could not collect

cc @malfet @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233 @Lezcano @albanD

**Comment by mruberry** [comment]
@albanD do we have an m1 oncall/label?

**Comment by mruberry** [comment]
I can confirm this doesn't reproduce on Colab with 1.9 nor on my devfair with a build built from trunk.

**Comment by albanD** [comment]
We don't have an oncall, but macos/arm label is the right thing to add here. I can try to repro locally.

@junhsss could you confirm how you installed pytorch on your machine?

**Comment by Lezcano** [comment]
cc @IvanYashchuk @malfet who recently were able to repro a similar issue arising on an M1.

**Comment by IvanYashchuk** [comment]
The same problem was reported in https://github.com/pytorch/pytorch/issues/60351 and it was fixed with https://github.com/pytorch/pytorch/pull/60703.
Currently, I don't have an access to M1, I can't verify that the reported issue with einsum is fixed.

**Comment by malfet** [comment]
Can confirm that issues reproduces in 1.9.0 release, but doesn't it nightly, which means that #60703 is probably the correct fix:
```
% python -c 'import torch;x=torch.tensor([1., 2.], dtype=torch.float32);torch.tensor([1., 2.], dtype=torch.float32); print(torch.__version__, x.dtype, ":", torch.einsum("i, i -> ", x, x)) '
1.9.0 torch.float32 : tensor(0.)
$ pip uninstall torch; pip install --pre torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
...
% python -c 'import torch;x=torch.tensor([1., 2.], dtype=torch.float32);torch.tensor([1., 2.], dtype=torch.float32); print(torch.__version__, x.dtype, ":", torch.einsum("i, i -> ", x, x)) '
1.10.0.dev20210728 torch.float32 : tensor(5.)
```
Closing as duplicate of #60351
